readthedocs/doc_builder/director.py

```diff
@@ -74,24 +74,33 @@ def setup_vcs(self):
             # ca-certificate package which is compatible with Lets Encrypt
             container_image=settings.RTD_DOCKER_BUILD_SETTINGS["os"]["ubuntu-20.04"],
         )
-        with environment:
+        with self.vcs_environment:
             before_vcs.send(
                 sender=self.data.version,
-                environment=environment,
+                environment=self.vcs_environment,
             )
 
             # Create the VCS repository where all the commands are going to be
             # executed for a particular VCS type
             self.vcs_repository = self.data.project.vcs_repo(
                 version=self.data.version.slug,
-                environment=environment,
+                environment=self.vcs_environment,
                 verbose_name=self.data.version.verbose_name,
                 version_type=self.data.version.type,
             )
 
-            self.pre_checkout()
+            # We can't do too much on ``pre_checkout`` because we haven't
+            # cloned the repository yet and we don't know what the user wrote
+            # in the `.readthedocs.yaml` yet.
+            #
+            # We could implement something different in the future if we download
+            # the `.readthedocs.yaml` file without cloning.
+            # See https://github.com/readthedocs/readthedocs.org/issues/8935
+            #
+            # self.run_build_job('pre_checkout')
+
             self.checkout()
-            self.post_checkout()
+            self.run_build_job("post_checkout")
```

Comment:
as mentioned before, post checkout will run on the container that does the clone step, which could be confusing. I think we decided to leave this one out for now.

Reply:
It's fine to run `post_checkout` in the same container. I don't think this will affect too much. I want this step, in particular for unshallow clones, so it can be controlled by the user.

I'd assume that all the commands executed on the `post_checkout` step will be associated with the repository itself, which is maintained between containers. We can also add a comment in the documentation about this to avoid confusion if any.